Correctly read the 'style' argument while processing 'genfrac'.

tfpf opened the pull request:
## PR Summary
An attempt to fix https://github.com/matplotlib/matplotlib/issues/23008.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

anntzer commented on lib/matplotlib/_mathtext.py:
```diff
@@ -2336,7 +2338,8 @@ def _genfrac(self, ldelim, rdelim, rule, style, num, den):
             toks.get("ldelim", ""), toks.get("rdelim", ""),
-            toks["rulesize"], toks["style"],
+            toks["rulesize"], self._MathStyle(
+                toks.get("style", self._MathStyle.TEXTSTYLE.value)),
```
I think that should be self._MathStyle.TEXTSTYLE (as the code downstream accesses the `.value` attribute itself)?

tfpf replied:
Was just about to post a comment about the same thing. This should work, because the whole thing is wrapped inside `self._MathStyle()`. So, an `enum` variable is correctly passed.

However, this does two conversions (`enum` to `int`, then back to `enum`). Is there a more optimal way to default to text style?

anntzer replied:
Ah sorry I missed that.  FWIW I would just turn the enum into an IntEnum and accept some confusion between ints and MathStyles; converting back and forth (while unlikely to cause a significant slowdown) doesn't seem to be really necessary or help legibility here.

tfpf commented:
Thought I'd try one more thing before `IntEnum`. Automatically converting the style code into an `enum` by making `style_literal` a separate function seems to be a better approach? (It solves the double conversion problem, too.)

anntzer commented:
Indeed, these are irrelevant.
Probably this could use a test using check_figures_equal comparing genfrac and dfrac?

tfpf commented:
Yes, I could add a test for that. One point, though: `\genfrac` needs `rulesize`, which only accepts floating-point numbers, whereas `\dfrac` uses a rule size of `self.get_state().get_current_underline_thickness()`. Not sure how to proceed.

If I set the rule size to 1 (e.g. `\genfrac{}{}{1}{0}{2x}{3y}` compared with `\dfrac{2x}{3y}`), the test passes (it fails for 0 and 2). But that doesn't look like a _good_ test?

tfpf commented:
> One point, though: `\genfrac` needs `rulesize`, which only accepts floating-point numbers, whereas `\dfrac` uses a rule size of `self.get_state().get_current_underline_thickness()`.

I am open to suggestions!

anntzer commented:
I agree this is a bit annoying :(  Probably I would cheat a bit and, in the test, use
```
thickness = matplotlib._mathtext.TruetypeFonts.get_underline_thickness(None, None, fontsize=<whatever>, dpi=<whatever>)
```
and insert that value into the test genfrac.

anntzer commented on lib/matplotlib/tests/test_mathtext.py:
```diff
@@ -343,6 +343,16 @@ def test_operator_space(fig_test, fig_ref):
+    fig_test.text(0.1, 0.1, r"$\dfrac{2x}{3y}$")
+
+    thickness = _mathtext.TruetypeFonts.get_underline_thickness(
+        None, None, fontsize=12, dpi=mpl.rcParamsDefault['figure.dpi'])
```
I think that should be rcParams, not rcParamsDefault (in theory, tests could configure a different rcParams dpi)

tfpf replied:
Ah, so that's what I was missing. Thanks for the heads-up.

anntzer commented on lib/matplotlib/tests/test_mathtext.py:
```diff
@@ -343,6 +343,16 @@ def test_operator_space(fig_test, fig_ref):
+    thickness = _mathtext.TruetypeFonts.get_underline_thickness(
+        None, None, fontsize=12, dpi=mpl.rcParamsDefault['figure.dpi'])
+    fig_ref.text(
+        0.1, 0.1, fr"$\genfrac{{}}{{}}{{{thickness}}}{{0}}{{2x}}{{3y}}$")
```
Probably a good place to use %-style formatting...

anntzer reviewed (approved):
Minor nits, which don't prevent the approval IMO.

tfpf commented:
Rebased on the current `main` branch, since a temporary fix for the failing macOS tests appears to be in place.

tfpf commented:
Fixed merge conflicts and rebased again. There are a couple of back-and-forth changes, so I will combine everything into a single commit.

The test failures are due to an unidentified image error. Not sure what went wrong.

oscargus commented:
I squashed the commit. Thanks!
